npc-utils.js

```diff
@@ -0,0 +1,570 @@
+
+    this.start.set(
+      Math.round(start.x),
+      start.y,
+      Math.round(start.z),
+    );
+    this.dest.set(
+      Math.round(dest.x),
+      dest.y,
+      Math.round(dest.z),
+    );
+
+    this.allowNearest = allowNearest;
+
+    this.startVoxel = this.createVoxel(this.start);
+    this.startVoxel._isStart = true;
+    this.startVoxel._isReached = true;
+    // this.startVoxel._priority = start.manhattanDistanceTo(dest)
+    this.startVoxel._priority = this.start.distanceTo(this.dest);
+    this.startVoxel._costSoFar = 0;
+    this.frontiers.push(this.startVoxel);
+
+    this.destVoxel = this.createVoxel(this.dest);
+    if (this.startVoxel === this.destVoxel) return;
+    this.destVoxel._isDest = true;
+
+    // // this.step();
+    this.untilFound();
+    if (this.isFound) {
+      this.simplifyWaypointResultXZ(this.waypointResult[0]);
+      this.simplifyWaypointResultXZ2(this.waypointResult[0]);
+      this.simplifyWaypointResultX(this.waypointResult[0]);
+      this.simplifyWaypointResultZ(this.waypointResult[0]);
+      this.waypointResult.shift();
+    }
+    // console.log('waypointResult', this.waypointResult.length);
+
+    if (this.debugRender) {
+      // const len = this.waypointResult.length - 1;
+      const len = this.waypointResult.length;
+      for (let i = 0; i < len; i++) {
+        const voxel = this.getVoxel(this.waypointResult[i].position);
+        if (voxel) { // May already disposed.
+          voxel._isPathSimplified = true;
+        }
+      }
+
+      this.debugMesh.count = this.voxels.children.length;
+      this.voxels.children.forEach((voxel, i) => {
+        this.debugMesh.setMatrixAt(i, voxel.matrix);
+        if (voxel._isStart) {
+          this.debugMesh.setColorAt(i, colorStart);
+        } else if (voxel._isDest) {
+          this.debugMesh.setColorAt(i, colorDest);
+        } else if (voxel._isPathSimplified) {
+          this.debugMesh.setColorAt(i, colorPathSimplified);
+        } else if (voxel._isPath) {
+          this.debugMesh.setColorAt(i, colorPath);
+        } else if (voxel._isFrontier) {
+          this.debugMesh.setColorAt(i, colorFrontier);
+        } else if (voxel._isReached) {
+          this.debugMesh.setColorAt(i, colorReached);
+        } else if (voxel._isIdle) {
+          this.debugMesh.setColorAt(i, colorIdle);
+        }
+      });
+      this.debugMesh.instanceMatrix.needsUpdate = true;
+      this.debugMesh.instanceColor.needsUpdate = true;
+    }
+
+    return this.isFound;
+  }
+
+  simplifyWaypointResultX(result) {
+    if (result?._next?._next) {
+      if (
+        result.position.x === result._next._next.position.x && // check whether in one line
+        Math.sign(result.position.z - result._next.position.z) === Math.sign(result._next.position.z - result._next._next.position.z) // check wheter in one layer // TODO: simplifyWaypointResultXZ() should need check wheter in one layer too.
```

Comment:
Unfinished todo.

Reply:
Yeah, it's a little hard and need a lot test, I'm keep working on this.
But I feel it's rare case or even will never appear due to the A* algorithm, I'll confirm.

Reply:
Done.
https://github.com/webaverse/app/pull/2298/commits/fdc03437cce000d9e8b4a2f6c964ed8b5f753e25
https://github.com/webaverse/app/pull/2298/commits/1d728d0d42736d829a384a3e28dec995be72cfa3